### PR TITLE
[FLINK-33983][runtime] Introduce JobEvent and JobEventStore

### PR DESCRIPTION
**Pull request by JunRuiLee**
## What is the purpose of the change

[FLINK-33983][runtime] Introduce JobEvent and JobEventStore

## Brief change log

Introduce JobEvent and JobEventStore

## Verifying this change

This change added tests and can be verified as follows:

- ExecutionJobVertexInitializedEventTest
- ExecutionVertexFinishedEventTest
- ExecutionVertexResetEventTest
- FileSystemJobEventStoreTest
- FsBatchFlushOutputStreamTest
- JobEventManagerTest
- TestingJobEvent

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+                    int binaryEventType = inputStream.readInt();
+                    int binaryEventSize = inputStream.readInt();
+                    byte[] binaryEvent = new byte[binaryEventSize];
+                    inputStream.readFully(binaryEvent);
+                    SimpleVersionedSerializer<JobEvent> serializer =
+                            checkNotNull(
+                                    jobEventSerializers.get(binaryEventType),
+                                    "There is no registered serializer for job event with type "
+                                            + binaryEventType);
+                    event = serializer.deserialize(1, binaryEvent);
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java:301]: 1 -> GenericJobEventSerializer#VERSION

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        return paths;
+    }
+
+    protected ScheduledExecutorService createJobEventWriterExecutor() {
+        return Executors.newSingleThreadScheduledExecutor(
+                new ExecutorThreadFactory("job-event-writer"));
+    }
+
+    @Override
+    public void stop(boolean clear) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java:191]: clear -> clearEventLogs

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+            if (clear) {
+                fileSystem.delete(workingDir, true);
+            }
+        } catch (Exception exception) {
+            LOG.warn("Fail to stop filesystem job event store.", exception);
+        }
+    }
+
+    void writeInt(FSDataOutputStream out, int num) throws IOException {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java:220]: This method can be part of `FsBatchFlushOutputStream`.
`FsBatchFlushOutputStream` can be explicit used in this store.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    @Override
+    public void writeEvent(JobEvent event, boolean cutBlock) {
+        checkNotNull(fileSystem);
+        checkNotNull(event);
+
+        eventWriterExecutor.execute(() -> writeEventRunnable(event, cutBlock));
+    }
+
+    @VisibleForTesting
+    protected void writeEventRunnable(JobEvent event, boolean cutBlock) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java:236]: Comments are needed to explain what does `cutBlock` mean.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/ExecutionVertexFinishedEvent.java [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                }
+
+                return jobEvent;
+            } catch (Exception exception) {
+                throw new IOException(
+                        "Deserialize ExecutionVertexFinishedEvent failed.", exception);
+            }
+        }
+
+        static byte[] readBytes(DataInputStream in, int size) throws IOException {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/ExecutionVertexFinishedEvent.java:239]: can be private

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/ExecutionVertexFinishedEvent.java [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        return "ExecutionVertexFinishedEvent("
+                + "executionVertexId='"
+                + getExecutionVertexId()
+                + "', attemptNumber='"
+                + getAttemptNumber()
+                + "')";
+    }
+
+    /** Serializer for {@link ExecutionVertexFinishedEvent}. */
+    public static class Serializer implements SimpleVersionedSerializer<JobEvent> {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/ExecutionVertexFinishedEvent.java:136]: can be package private

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+                    byte[] binaryEvent = new byte[binaryEventSize];
+                    inputStream.readFully(binaryEvent);
+                    SimpleVersionedSerializer<JobEvent> serializer =
+                            checkNotNull(
+                                    jobEventSerializers.get(binaryEventType),
+                                    "There is no registered serializer for job event with type "
+                                            + binaryEventType);
+                    event = serializer.deserialize(1, binaryEvent);
+                } catch (EOFException eof) {
+                    closeInputStream();
```
[the remark this answers is not in the document]
Comment (JunRuiLee) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStore.java:303]: Yes, DataInputStream lacks a method to precisely indicate the end of a file. The available() method's output isn't reliable for determining file completion, as it doesn't always provide an accurate count of readable bytes left. 

So handling EOFException is a common yet inelegant solution for detecting the end of a file.

**Comment by JunRuiLee** [comment]
Thanks @zhuzhurk for review, I've updated this pr accordingly. PTAL.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStoreTest.java [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+                            try {
+                                store.getOutputStream().writeInt(TestingJobEvent.TYPE_ID);
+                                store.getOutputStream().writeInt(5000);
+                                store.writeEvent(new TestingJobEvent(100));
+                            } catch (IOException e) {
+                                exception.set(e);
+                            }
+                        })
+                .get();
+        assertThat(exception.get())
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStoreTest.java:214]: What's this assertion for?

**Review by zhuzhurk — COMMENTED** [review]
Thanks for addressing all the comments! @JunRuiLee 
I have last two minor comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FsBatchFlushOutputStream.java [hunk 9]
```diff
@@ -107,9 +114,6 @@ public void flush() throws IOException {
                 pos = 0;
             }
         }
-        if (LOG.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (JunRuiLee) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/event/FsBatchFlushOutputStream.java]: updated.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStoreTest.java (hunk 8 above)
Comment (JunRuiLee) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/event/FileSystemJobEventStoreTest.java]: removed.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request merged**